### PR TITLE
V2 index for string and URI

### PR DESCRIPTION
**Pull request by michaelabuckley**
Change string and URI indexes to handle exact match faster.
Change :above query on URI to use hash_identity instead of old sp_name query.

### CODE REVIEW

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4StandardQueriesNoFTTest.java [hunk 1]
```diff
@@ -73,6 +77,87 @@ protected FhirContext getFhirContext() {
...
+			"normal search matches lower prefix , Flintstones,  =flints     ,  True",
+			"normal search matches mixed prefix , Flintstones,  =fLiNtS     ,  True",
+			"normal search ignores accents      , Flíntstones,  =Flintstone ,  True",
+			"normal search no match suffix      , Flintstones,  =intstones  ,  False",
+			"normal search matches first letter , Flintstones,  =f          ,  True",
+			"exact  search matches exact        , Flintstones,  :exact=Flintstones ,  True",
+			"exact  search no match wrong case  , Flintstones,  :exact=flintstones ,  False",
+			"exact  search no match prefix      , Flintstones,  :exact=Flint       ,  False",
+//			"contains search match prefix      ,  Flintstones,  :contains=flint       ,  True",
+//			"contains search match prefix      ,  Flintstones,  :contains=Flint       ,  True",
```
Comment (jmarchionatto) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4StandardQueriesNoFTTest.java:96]: No more testing for these cases?

Comment (michaelabuckley) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4StandardQueriesNoFTTest.java:96]: We don't support :contains without Lucene.

**Comment by patrick-vachon-smilecdr** [comment]
Looks good to me!

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java [hunk 2]
```diff
@@ -201,12 +201,58 @@ protected void init660() {
...
+			// string search index
+			Builder.BuilderWithTableName stringTable = version.onTable("HFJ_SPIDX_STRING");
+
+			// add res_id to indentity to speed up sorts.
+			stringTable
+				.addIndex("20230303.1", "IDX_SP_STRING_HASH_IDENT_V2")
+				.unique(false)
+				.online(true)
+				.withColumns("HASH_IDENTITY", "RES_ID", "PARTITION_ID");
+			stringTable.dropIndexOnline("20230303.2", "IDX_SP_STRING_HASH_IDENT");
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java:222]: is this synchronous?  could it slow down startup time?

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IResourceIndexedSearchParamUriDao.java [hunk 3]
```diff
@@ -30,8 +30,8 @@
 
 public interface IResourceIndexedSearchParamUriDao extends JpaRepository<ResourceIndexedSearchParamUri, Long>, IHapiFhirJpaRepository {
 	
-	@Query("SELECT DISTINCT p.myUri FROM ResourceIndexedSearchParamUri p WHERE p.myResourceType = :resource_type AND p.myParamName = :param_name")
-	public Collection<String> findAllByResourceTypeAndParamName(@Param("resource_type") String theResourceType, @Param("param_name") String theParamName);
+	@Query("SELECT DISTINCT p.myUri FROM ResourceIndexedSearchParamUri p WHERE p.myHashIdentity = :hash_identity")
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IResourceIndexedSearchParamUriDao.java:33]: will this work on partitioned servers?

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java (hunk 2 above)
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java:222]: is this synchronous? could it slow down startup time?

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/predicate/UriPredicateBuilder.java [hunk 4]
```diff
@@ -113,7 +114,8 @@ public Condition addPredicate(List<? extends IQueryParameterType> theUriOrParame
 						.add(StorageProcessingMessage.class, message);
 					CompositeInterceptorBroadcaster.doCallHooks(myInterceptorBroadcaster, theRequestDetails, Pointcut.JPA_PERFTRACE_WARNING, params);
 
-					Collection<String> candidates = myResourceIndexedSearchParamUriDao.findAllByResourceTypeAndParamName(getResourceType(), theParamName);
+					long hashIdentity = BaseResourceIndexedSearchParam.calculateHashIdentity(getPartitionSettings(), getRequestPartitionId(), getResourceType(), theParamName);
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/predicate/UriPredicateBuilder.java:117]: oh it looks like it is partition safe...

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IResourceIndexedSearchParamUriDao.java (hunk 3 above)
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IResourceIndexedSearchParamUriDao.java:33]: Yes.  Partition is part of hash_identity.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java (hunk 2 above)
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java:222]: The create index?  Yes.

Comment (michaelabuckley) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java]: Added upgrade.md notes.

**Pull request merged**